ratelimit/src/lib.rs

```diff
@@ -1,203 +1,470 @@
-                Ok(Refill::Uniform) => self.uniform.sample(&mut rand::thread_rng()) as u64,
-                Ok(Refill::Normal) => self.normal.sample(&mut rand::thread_rng()) as u64,
-                Err(_) => self.tick.load(Ordering::Relaxed).as_nanos(),
-            };
-            if self
-                .next
-                .compare_exchange(
-                    next,
-                    next + Duration::<Nanoseconds<u64>>::from_nanos(tick),
-                    Ordering::SeqCst,
-                    Ordering::SeqCst,
-                )
-                .is_ok()
-            {
-                let quantum = self.quantum.load(Ordering::Relaxed);
-                let capacity = self.capacity.load(Ordering::Relaxed);
-                let available = self.available.load(Ordering::Relaxed);
-                if available + quantum >= capacity {
-                    let quantum = capacity - available;
-                    self.available.fetch_add(quantum, Ordering::Relaxed);
-                } else {
-                    self.available.fetch_add(quantum, Ordering::Relaxed);
+    /// Consumes this `Builder` and produces a `Ratelimiter`.
+    pub fn build(self) -> Ratelimiter {
+        let (amount, interval) = match (self.refill_amount, self.refill_interval) {
+            (Some(amount), Some(interval)) => {
+                // if we have an explicit amount and interval, use those
+                // directly
+                (amount, interval.as_nanos() as u64)
+            }
+            (Some(amount), None) => {
+                let a = self.refill_rate.0;
+                let i = self.refill_rate.1.as_nanos() as u64;
+
+                // tokens per nanosecond
+                let target_rate = a as f64 / i as f64;
+
+                let interval = (amount as f64 / target_rate).round() as u64;
+
+                (amount, interval)
+            }
+            (None, Some(interval)) => {
+                let a = self.refill_rate.0;
+                let i = self.refill_rate.1.as_nanos() as u64;
+
+                // tokens per nanosecond
+                let target_rate = a as f64 / i as f64;
+
+                let amount = (interval.as_nanos() as u64 as f64 * target_rate).round() as u64;
+                let interval = interval.as_nanos() as u64;
+
+                (amount, interval)
+            }
+            (None, None) => {
+                // work out an amount and interval based on the rate
+
+                let mut amount = self.refill_rate.0;
+                let mut interval = self.refill_rate.1.as_nanos() as u64;
+
+                // copy the original values for use later
+                let a = amount;
+                let i = interval;
+
+                // reduce things to the smallest amount and interval possible
+                let gcd = gcd(amount, interval);
+                amount /= gcd;
+                interval /= gcd;
+
+                // scale things up to reduce bookeeping
+                while interval < 100 {
```

Comment:
I prefer a crate-level constant like `MIN_INTERVAL` to make this explicit.

Reply:
side-stepped by avoiding rate -> amount / interval calculations in the crate